library/src/image.cxx

```diff
@@ -0,0 +1,194 @@
+image& image::operator=(const image& img)
+{
+  *this->Internals = *img.Internals;
+  return *this;
+}
+
+//----------------------------------------------------------------------------
+image::image(image&& img)
+{
+  this->Internals = img.Internals;
+  img.Internals = nullptr;
+}
+
+//----------------------------------------------------------------------------
+image& image::operator=(image&& img)
+{
+  std::swap(this->Internals, img.Internals);
+  return *this;
+}
+
+//----------------------------------------------------------------------------
+int image::getWidth() const
+{
+  return this->Internals->Width;
+}
+
+//----------------------------------------------------------------------------
+int image::getHeight() const
+{
+  return this->Internals->Height;
+}
+
+//----------------------------------------------------------------------------
+image& image::setResolution(int width, int height)
+{
+  this->Internals->Width = width;
+  this->Internals->Height = height;
+  this->Internals->UpdateBufferSize();
+  return *this;
+}
+
+//----------------------------------------------------------------------------
+int image::getChannelCount() const
+{
+  return this->Internals->Channels;
+}
+
+//----------------------------------------------------------------------------
+image& image::setChannelCount(int dim)
+{
+  this->Internals->Channels = dim;
+  this->Internals->UpdateBufferSize();
+  return *this;
+}
+
+//----------------------------------------------------------------------------
+image& image::setData(unsigned char* buffer)
+{
+  std::copy(this->Internals->Buffer.begin(), this->Internals->Buffer.end(), buffer);
+  return *this;
+}
+
+//----------------------------------------------------------------------------
+unsigned char* image::getData() const
+{
+  return this->Internals->Buffer.data();
+}
+
+//----------------------------------------------------------------------------
+bool image::compare(const image& reference, image& result, double threshold) const
+{
+  auto importerThis = this->Internals->GetVTKImporter();
+  auto importerRef = reference.Internals->GetVTKImporter();
+
+  vtkNew<vtkImageDifference> diff;
+  // handle threshold outside of vtkImageDifference:
+  // https://gitlab.kitware.com/vtk/vtk/-/issues/18152
+  diff->SetThreshold(0);
+  diff->SetInputConnection(importerThis->GetOutputPort());
+  diff->SetImageConnection(importerRef->GetOutputPort());
+  diff->UpdateInformation();
+  double error = diff->GetThresholdedError();
+
+  if (error <= threshold)
+  {
+    diff->Update();
+    error = diff->GetThresholdedError();
+  }
+
+  result.Internals->SetFromVTK(diff);
+
+  return error <= threshold;
+}
+
+//----------------------------------------------------------------------------
+image& image::save(const std::string& path)
+{
+  auto importer = this->Internals->GetVTKImporter();
+
+  vtkNew<vtkPNGWriter> writer;
+  writer->SetFileName(path.c_str());
```

Comment:
check path is valid and can be written on ?

Reply:
This one is quite hard. How to determine if a path is writable?

Reply:
right